worlds/sc2/__init__.py

```diff
@@ -160,6 +168,26 @@ def create_items(self):
 
         self.multiworld.itempool += pool
 
+        # Tell the logic which unit classes are used for required W/A upgrades
+        used_item_names: Set[str] = {item.name for item in item_list}
+        used_item_names = used_item_names.union(item.name for item in self.multiworld.itempool if item.player == self.player)
+        if used_item_names.isdisjoint(item_groups.barracks_units):
+            self.has_barracks_unit = False
+        if used_item_names.isdisjoint(item_groups.factory_units):
+            self.has_factory_unit = False
+        if used_item_names.isdisjoint(item_groups.starport_units):
+            self.has_starport_unit = False
```

Comment:
These checks have minor holes in them, particularly for attack upgrades for casters or units that get their damage from abilities. So that would be:
* Widow Mine
* Raven
* Science Vessel

And possibly others I'm not thinking of.

I notice in the item groups changes, you left some things out for protoss if there was no attack, particularly observers and havocs. If we were trying to be consistent here, that would mean the above units shouldn't force either attack or armour checks into the logic.

Kind of a pain to deal with and a super-minor edge case, so maybe just leave a comment around here noting this minor issue for anyone looking at this code in future and pointing out the inconsistency. Just so if a crazy yaml engineer down the line wants to know why their widow mine / raven-only yaml is requiring attack upgrades, they just have to spot the comment rather than re-engineer this whole system.

Reply:
Solved